pkg/util/quotapool/intpool_test.go

```diff
@@ -0,0 +1,456 @@
+			if _, isErrClosed := err.(*quotapool.ErrClosed); !isErrClosed {
+				t.Fatal(err)
+			}
+		}
+	}
+
+	go tryAcquire()
+
+	select {
+	case <-time.After(5 * time.Second):
+		t.Fatal("quota pool closing did not unblock acquisitions within 5s")
+	case err := <-resCh:
+		if _, isErrClosed := err.(*quotapool.ErrClosed); !isErrClosed {
+			t.Fatal(err)
+		}
+	}
+}
+
+// TestQuotaPoolCanceledAcquisitions tests the behavior where we enqueue
+// multiple acquisitions with canceled contexts and expect any subsequent
+// acquisition with a valid context to proceed without error.
+func TestQuotaPoolCanceledAcquisitions(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	qp := quotapool.NewIntPool("test", 1)
+	alloc, err := qp.Acquire(ctx, 1)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	cancel()
+	const numGoroutines = 5
+
+	errCh := make(chan error)
+	for i := 0; i < numGoroutines; i++ {
+		go func() {
+			_, err := qp.Acquire(ctx, 1)
+			errCh <- err
+		}()
+	}
+
+	for i := 0; i < numGoroutines; i++ {
+		select {
+		case <-time.After(5 * time.Second):
+			t.Fatal("context cancellations did not unblock acquisitions within 5s")
+		case err := <-errCh:
+			if err != context.Canceled {
+				t.Fatalf("expected context cancellation error, got %v", err)
+			}
+		}
+	}
+
+	alloc.Release()
+	go func() {
+		_, err := qp.Acquire(context.Background(), 1)
+		errCh <- err
+	}()
+
+	select {
+	case err := <-errCh:
+		if err != nil {
+			t.Fatal(err)
+		}
+	case <-time.After(5 * time.Second):
+		t.Fatal("acquisition didn't go through within 5s")
+	}
+}
+
+// TestQuotaPoolNoops tests that quota pool operations that should be noops are
+// so, e.g. quotaPool.acquire(0) and quotaPool.release(0).
+func TestQuotaPoolNoops(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	qp := quotapool.NewIntPool("test", 1)
+	ctx := context.Background()
+	initialAlloc, err := qp.Acquire(ctx, 1)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Acquisition of blockedAlloc will block until initialAlloc is released.
+	errCh := make(chan error)
+	var blockedAlloc *quotapool.IntAlloc
+	go func() {
+		blockedAlloc, err = qp.Acquire(ctx, 1)
+		errCh <- err
+	}()
+
+	// Allocation of zero should not block.
+	emptyAlloc, err := qp.Acquire(ctx, 0)
+	if err != nil {
+		t.Fatalf("failed to acquire 0 quota: %v", err)
+	}
+	emptyAlloc.Release() // Release of 0 should do nothing
+
+	initialAlloc.Release()
+	select {
+	case <-time.After(5 * time.Second):
+		t.Fatal("context cancellations did not unblock acquisitions within 5s")
+	case err := <-errCh:
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+	if q := qp.ApproximateQuota(); q != 0 {
+		t.Fatalf("expected quota: 0, got: %d", q)
+	}
+	blockedAlloc.Release()
+	if q := qp.ApproximateQuota(); q != 1 {
+		t.Fatalf("expected quota: 1, got: %d", q)
+	}
+}
+
+// TestQuotaPoolMaxQuota tests that at no point does the total quota available
+// in the pool exceed the maximum amount the pool was initialized with.
```

Comment:
Hmm.. this test looks very similar to the first one in this file. Where is it doing what this claims?